src/structs.js

```diff
@@ -0,0 +1,459 @@
+});
+
+const {
+    LairClientError,
+    HEADER_SIZE,
+}					= require('./constants.js');
+const {
+    LairType,
+    LairString,
+    LairSized,
+    LairPublicKey,
+    LairSignature,
+    LairKeystoreIndex,
+    LairEntryType,
+    LairDigest,
+    LairCert,
+    LairCertSNI,
+    LairCertAlgorithm,
+    LairCertPrivateKey,
+    ConversionError,
+}					= require('./types.js');
+
+
+class LairStruct {
+    [Symbol.toStringTag]		= LairStruct.name;
+
+    static from ( source, has_header_prefix = false ) {
+	source				= Buffer.from( source );
+	let struct			= new this();
+
+	if ( has_header_prefix === true ) {
+	    log.info("Skipping header (%s) bytes in source: %s", () => [ HEADER_SIZE, source.slice(0, HEADER_SIZE).toString("hex") ]);
+	    source			= source.slice( HEADER_SIZE );
+	}
+
+	for (let i=0; i < this.PAYLOAD.length; i++) {
+	    let type			= this.PAYLOAD[i];
+	    log.silly("Extracting type '%s' from source (%s): %s", () => [
+		type.name, source.length, source.toString("hex") ]);
+	    let value			= type.fromSource( source, false );
+	    source			= source.slice( value.length );
+
+	    struct.push( value );
+	}
+
+	if ( source.length > 0 )
+	    log.warn("source (%s) for struct %s still has some bytes leftover; this may be unexpected.", source.length, this.name );
+
+	return struct;
+    }
+
+    constructor ( ...args ) {
+	this.max_length			= this.constructor.PAYLOAD.length;
+
+	if ( this.max_length < args.length )
+	    throw new ConversionError(`Received more arguments (${args.length}) than template allows for ${this.name}`);
+
+	this._values			= new Array(this.max_length);
+	this.constructor.PAYLOAD.map( (type, i) => {
+	    Object.defineProperty( this, i, {
+		get () {
+		    return this._values[i];
+		},
+		set ( value ) {
+		    if ( this._values[i] !== undefined )
+			this.byteLength -= this._values[i].length;
+
+		    if ( !(value instanceof LairType) )
+			value		= type.from( value );
+
+		    log.silly("Setting '%s' (%s bytes) to satisfy PAYLOAD item %s (%s)", () => [
+			i, value.length, type.name, type.SIZE ]);
+		    this._values[i]	= value;
+		    this.byteLength    += value.length;
+		},
+	    });
+	});
+
+	this.byteLength			= HEADER_SIZE;
+	args.map( value => this.push(value) );
+    }
+
+    push ( value ) {
+	let i				= this._values.filter(Boolean).length;
+
+	if ( i === this.max_length )
+	    throw new Error(`Cannot push another item onto ${this.constructor.name} because it already contains the expected number of payoad items.`);
+
+	log.silly("Pushing '%s' (%s) to satisfy PAYLOAD item %s", () => [
+	    value.constructor.name, typeof value, i ]);
+	this[i]				= value;
+
+	return i + 1;
+    }
+
+    toMessage ( msg_id ) {
+	if (msg_id === undefined) {
+	    if ( this instanceof LairResponse )
+		throw new TypeError(`Message ID must be provided for Lair responses`);
+	    else
+		msg_id			= Math.floor(Math.random() * 2**32);
```

Comment:
While a handful in 4 billion-ish chance of collision is pretty low, you could also have an auto-incrementing message ID that would eliminate the chance entirely : )

Reply:
Good catch, I had a mental note to update this but obviously forgot.  I'll fix this before the release and then hit resolve.

Reply:
Fixed in https://github.com/Holo-Host/lair-client-js/pull/1/commits/b66a0a81a7b32bcc5ab2e98c933bd0eab2c97848